README
YACA supported instructions:
(Yet Another (nasty) Chip8 Assembler)

- Support of Chip8/SuperChip8 instruction
- ADD
- LD
- CLS
- AND
- OR
- XOR
- SUB
- SUBN
- JMP/JP
- CALL
- SCD
- SCR
- SCL
- LOW
- HIGH
- RET
- SE
- SNE
- SHR
- RSB
- SHL
- MVI
- JMI
- RND
- DRW
- SKP
- SKNP
- SYS
- EXIT
- DB
- DW

Example directory:
 - Minimal example proposed in the compiler pack from rivial studios (http://www.revival-studios.com). 
 - Only a test prog, it does not do anything.

CHANGELOG
TODO:
- Add preproc. keywords
- yaca.c and code.c refactoring
- Improve syntax error management
- Add multi data definition

FIXED:
- Modifying some mnemonic name
- Added ';' as comment beginning
- Fixed many bugs
- Added "db" and "dw" keywords



examples/test.asm
	CLS
	CALL lab2
	SCD #4
	LD v8, #0xFFF
loop5:
	SCR
	SCL
	SE Va,v6
	JMP loop1
	ADD VF,Vf
	ADD V4,#0b100
	LD V4,v5

loop2:	ADD v3,v2
	SNE v3,#0
	OR Vf,Ve
	AND VE,VE
	XOR VE,VF
	JMP loop2
	SUB VA,va
	SUBN Va,VA
	SHR VA
	SHR VA,VB
	LD I,lab1
	SHL VB,VB
	JMP V0, lab1
	RND  V5, #0xFF	
	DRW  V1,v2,#3
	SKNP  VE
	SKP  vb
	LD   vc, DT
	LD Va,K
	LD DT,Vb
	LD     st,VC
	ADD    i, VF
	LD     f, V3
	ld     b, Ve
	ld [i], V3
	ld V3,[i]
	EXIT 

lab1:
	RET
lab2:
	RET

lab3:	DB	1

yaca.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "yaca.h"
#include "code.h"


int _debug = 0;
extern int error;

static prog mprog;
static lex *lab[50] = {NULL};
static char *insts[] = {
  "cls", 
  "ret", 
  "add", 
  "ld",
  "and",
  "or",
  "xor",
  "sub",
  "subn",
  "jmp",
  "jp",
  "call",
  "scd",
  "scr",
  "scl",
  "low",
  "high",
  "se",
  "sne",
  "shr",
  "rsb",
  "shl",
  "mvi",
  "jmi",
  "rnd",
  "drw",
  /*  "xsprite" ,*/
  "skp",
  "sknp",
  "sys",
  "exit",
  "db",
  "dw",
  NULL
};

static int change_val_rep = 1;

#define STR2LOW(s)				\
  {						\
    char *tmp = s;				\
    while(*tmp != '\0'){			\
      *tmp = tolower(*tmp);			\
      tmp++;					\
    }						\
  }

#define RM_CR(s)					\
  {							\
    char *tmp = s+strlen(s);				\
    while(tmp != s && *tmp != '\15'){			\
      tmp--;						\
    }							\
    if (*tmp == '\15') *tmp = '\0';			\
  }

#define CHAR2VAL(s)  (int)(s == '.'?0:((s >= 'a' && s <= 'f')?s-'a'+10:s-'0'))

void populate_labels(void)
{
  expr *tmp;
  int i = 0;
  for (tmp=mprog.first; tmp != NULL; tmp = tmp->next)
    {
      if (tmp->label != NULL && tmp->label->kind == LAB)
	lab[i++] = tmp->label; 
    }
}

void print_pop_labs(void)
{
  int i;
  
  printf("Labels\n");
  for (i = 0; i<50 && lab[i] != NULL ; i++)
    printf("\t%s, 0x%X\n", lab[i]->name, lab[i]->val);

}

/* Prog and structures management */
void init_prog(void)
{
  mprog.last = NULL;
  mprog.first = mprog.last;
  mprog.lines = 0;
}

lex * create_lex(char *name, int val, enum lex_kind kind)
{
  lex *l = (lex *)malloc(sizeof(lex));

  l->name = name;
  l->val = val;
  l->kind = kind;

  return l;
}

expr * create_expr(int addr, lex *label, lex *inst, lex *src, lex *dest, lex *op, int line)
{
  expr *e = (expr *)malloc(sizeof(expr));
 
  e->addr = addr;
  e->label = label;
  e->inst = inst;
  e->src = src;
  e->dest = dest;
  e->op = op;
  e->line = line; 
  e->next = NULL;

  return e;
}

void add_expr(expr *e)
{
  if (mprog.first == NULL){
    mprog.first = e;
    mprog.last = e;
  }
  else {
    mprog.last->next = e;
    mprog.last = mprog.last->next;
  }
  mprog.lines++;
}

void free_lex(lex *l)
{
  if (l == NULL) return;
  if (l->name != NULL) free(l->name);
  free(l);
}

void free_expr(expr *e)
{
  if (e == NULL) return;
  free_lex(e->label);
  free_lex(e->inst);
  free_lex(e->src);
  free_lex(e->dest);
  free(e);
}

void free_prog()
{
  expr *tmp;
  
  /* CAUTION */
  for (tmp=mprog.first; tmp != NULL; tmp = tmp->next)
    free_expr(tmp);
}

void print_lex(lex *l)
{
  printf("Type=%d, name=%s, val=%d\n", l->kind, l->name, l->val);
}

void print_expr(expr *e)
{
  printf("Addr=0x%x\n", e->addr);
  if (e->label != NULL) print_lex(e->label);
  if (e->inst != NULL) print_lex(e->inst);
  if (e->src != NULL) print_lex(e->src);
  if (e->dest != NULL) print_lex(e->dest);
}

void print_prog()
{
  expr *tmp;
  printf("Lines %d\n",mprog.lines);
  for (tmp=mprog.first; tmp != NULL; tmp = tmp->next)
    {
      print_expr(tmp);
    }  
}


/* Manage tokens */
lex * create_lab(char *s, int addr)
{
  char *tmp;
  tmp = malloc(sizeof(char)*(strlen(s)-1));
  s[strlen(s)-1]='\0';
  STR2LOW(s);
  strcpy(tmp,s);

  return create_lex(tmp,addr,LAB);
}

lex * create_mem(char *s)
{
  char *tmp;
  tmp = malloc(sizeof(char)*(strlen(s)));
  strcpy(tmp,s);

  return create_lex(tmp,0,ADDR);
}

lex * create_inst(char *s)
{
  char *tmp;
  tmp = malloc(sizeof(char)*strlen(s));
  strcpy(tmp,s);
  return create_lex(tmp,0,INST);
}

lex * create_reg(char *s)
{
  char *tmp;
  int num = (int)((s[1] >= 'a' && s[1] <= 'f')?s[1]-'a'+10:s[1]-'0');
  tmp = malloc(sizeof(char)*sizeof(s));
  strcpy(tmp,s);
  return create_lex(tmp,num,REG);
}

lex * create_val(char *s)
{
  char *tmp;
  int val = 0;
  int dec = 0;

  tmp = malloc(sizeof(char)*strlen(s));
  if (change_val_rep) {
    strcpy(tmp,s);
    if (!strncmp(s,"$",1)) dec = 1, s++;
    if (!strncmp(s,"#",1)) dec = 4, s++;
  }
  else {
    strcpy(tmp,++s);
    STR2LOW(s);
    if (!strncmp(s,"0b",2)) dec = 1, s+=2;
    if (!strncmp(s,"0x",2)) dec = 4, s+=2;
  }

  if (dec != 0) {
    val =  CHAR2VAL(*s);
    for (s++;*s!='\0'; s++)
      val = (val<<dec) | CHAR2VAL(*s); 
  } else {
    val = CHAR2VAL(*s);
    for (s++;*s!='\0'; s++)
      val = val*10+CHAR2VAL(*s);
  } 
  return create_lex(tmp,val,VAL);
}

int is_inst(char *s)
{
  int i = 0;

  STR2LOW(s);  
  while (insts[i] != NULL && strcmp(insts[i], s))
    i++;
  
  return (insts[i] != NULL);
}

int is_label(char *s)
{
  return (s[strlen(s)-1] == ':');
}

int is_reg(char *s)
{  
  STR2LOW(s);
  
  return (strlen(s) == 2 
	  && s[0] == 'v' 
	  && isxdigit(s[1]));
}

int is_new_val(char *s)
{
  if (*s != '#') return (0==1);
  s++;
  
  
  while (*s != '\0'){
    if (*s != 'X' && *s != 'x' 
	&& !isxdigit(*s)) return (0==1); 
    s++;
  }
  
  return (0==0);
}


int is_old_val(char *s)
{
  
  while (*s != '\0'){
    if (*s != '#' && *s != '$' && *s != '.' 
	&& !isxdigit(*s)) return (0==1); 
    s++;
  }
  
  return (0==0);
}

int is_val(char *s)
{
  if (change_val_rep) return is_old_val(s);
  else return is_new_val(s);    
}

int is_white_line(char *s)
{
  int l = strlen(s), res = 0;
  
  for (;*s!='\0'; s++)
    if (*s == ' ' || *s == '\t' || *s == '\n' || *s == '\15')
      res++;
  
  return (res==l);
}

void print_help()
{
  printf("YACA: Yet Another (nasty) Chip8 Compiler\n");
  printf("Usage: yaca {-a -o} [asm_file] [output_name]\n");
  printf("  --other, -o  \t manage other int representation\n  \t\t  ");
  printf("(with a '#' before each int and '0x' and '0b' \n\t\t  for hex and bin values)\n");
  printf("  --addr, -a n\t change base adress to n \n\t\t (in hex format 0xYYY, n could not be higher than 0xFFF)\n ");
}

int main(int argc, char *argv[])
{
  FILE *fin, *fout;
  char delims[] = "\t, ";
  char *result = NULL;
  char *str = (char *)malloc(1024 * sizeof(char));
  int base_addr = DEFAULT_ADDR;
  expr *first;
  int sd;
  int next = 1;
  int line = 1;

  /* Testing command line arguments */
  if (!strcmp(argv[1],"-h") 
      || !strcmp(argv[1],"--help")) print_help(), exit(0);

  if (argc < 3)
    {
      fprintf(stderr,"Usage: yaca {-o} [asm_file] [output_name]\n");
      exit(BAD_ARGS);
    }
  
  for (;next<argc-2;next++)
    if (!strcmp(argv[next],"-v")) _debug = 1;
    else if (!strcmp(argv[next],"--other") 
	     || !strcmp(argv[next], "-o"))  change_val_rep = 0;
    else if (!strcmp(argv[next],"--addr") 
	     || !strcmp(argv[next], "-a")){
      STR2LOW(argv[++next]);
      if (strncmp(argv[next],"0x",2)) 
	fprintf(stderr, "Bad addr format\n"), exit(BAD_ARGS);
      argv[next]+=2;
      base_addr =  CHAR2VAL(*argv[next]);
      for (*argv[next]++;*argv[next]!='\0'; argv[next]++)
	base_addr = (base_addr<<4) | CHAR2VAL(*argv[next]); 
    }
    
  
  /* Opening the input asm file */
      if ((fin=fopen(argv[next],"r")) == NULL)
	{
	  fprintf(stderr,"There is no file: %s\n", argv[next]);
	  exit(BAD_FILE);
	}

      /* Manage parsing and tokens */
      while (!feof(fin)){
	lex *lab = NULL, *inst = NULL, *srcdest[3];
	char copy[1024];
	int ops, i;
         
	fgets(str, 1023, fin); /* Get a line */
	RM_CR(str);
	while (str[0] == '\0' && !strcmp(str,"\n") && !feof(fin)){
	  fgets(str, 1023, fin); /* Get anoter line */
	  RM_CR(str);
	  line++;
	  if (_debug) printf("We skip blank lines\n");
	}
	if (feof(fin) && strlen(str) == 0) break; /* Ugly hack */

	for (i=0; str[i]!='\n' && str[i] != '\0' ; i++) ;
	str[i] = '\0';
    
	if (_debug) strcpy(copy,str); /* copy the line for display purposes */

	result = strtok( str, delims ); /* Parse into tokens */
	ops = 0;
	sd = 0; /* Another ugly hack */
	srcdest[0] = NULL; /* Again */
	srcdest[1] = NULL; /* Again */
	srcdest[2] = NULL; /* Again */
	while( result != NULL ) {
	  if (ops > 5){
	    fprintf(stderr,"Warning - Line %d: I do not manage multiple mem zone yet\n", line); 
	    ops = 0;
	    break;
	  }
	  if (*result == ';') break;
	  if (is_white_line(result)) break;
	  if (is_label(result)) lab = create_lab(result,base_addr);
	  else if (is_inst(result)) inst = create_inst(result);
	  else if (is_reg(result)) srcdest[sd++] = create_reg(result);
	  else if (is_val(result)) srcdest[sd++] = create_val(result);
	  else srcdest[sd++] = create_mem(result);
	  result = strtok( NULL, delims );
	  ops++;
	}
        
	if (ops > 0)
	  {
	    if (_debug) printf("LINE %d\n", line);
	    if (_debug) printf("On line %s\n there is %d ops\n", copy, ops);
	    if (ops >= 3 && lab == NULL)
	      first = create_expr(base_addr,lab,inst,srcdest[1],srcdest[0],srcdest[2],line);
	    else 
	      first = create_expr(base_addr,lab,inst,srcdest[0],srcdest[1],srcdest[2],line);
	    add_expr(first);
	    if (inst != NULL){
	      if (strcmp(inst->name,"db")) base_addr += 2;
	      else base_addr++;
	    }
	  }
	line++;
      }

      if (_debug) print_prog();
 
      populate_labels();
 
      if (_debug) print_pop_labs();
  
      /* Output file opening */
      if ((fout=fopen(argv[next+1], "w+")) == NULL){
	fprintf(stderr,"Error while creating %s\n", argv[next+1]);
	exit(BAD_FILE);
      }

      code_prog(mprog,lab,fout);

      free_prog();
      free(str);
      fclose(fin);
      fclose(fout);

      if (error){
	fprintf(stderr,"%d errors found!\n", error);
	remove(argv[next+1]);
      }

      return 0;
    }


examples/minimal.asm
; ---- minimal chip8 game ---
 

; V0-2 	scrap
; V3 	player x
; V4 	player y
; V5	enemy x
; V6	enemy y
; VF    collision detection

; keys:
; #2 = up
; #4 = left
; #5 = space
; #6 = right
; #8 = down
; #A = Ctrl
; #B = shift
; #C = Z
; #F = Space


;HIGH		; Uncomment this for Hires Superchip mode (128x64)

GAMEINIT:

CLS		; clear screen
CALL INITPLAYER ; init player 
CALL INITENEMY  ; init enemy 

CALL DRAWPLAYER ; draw player on startpos
CALL DRAWENEMY  ; draw enemy on startpos


; ================== game loop ================
GAMELOOP:	
	
  CALL DRAWPLAYER   ; draw xor player (clear last pos)
  CALL HANDLEINPUT  ; move player	
  CALL DRAWPLAYER   ; draw player on new pos

  CALL DRAWENEMY
  CALL MOVEENEMY
  CALL DRAWENEMY  

  ; check collision, init player upon collision
  SNE VF,#1
  JMP GAMEINIT

  LD V0,#1
  LD DT,V0
  CALL WAITDELAY
  

JP GAMELOOP

; ================== functions ================
INITPLAYER:
  LD V3,#32
  LD V4,#25
  RET

; drawplayer  (uses: v3,v4,I)
DRAWPLAYER:
  LD I,SPR_DUDE	   ;Set I to SPR_DUDE adress
  DRW V3,V4,#6
  RET

; handle input (uses v0)
HANDLEINPUT:
  ; check down
  LD V0,#8	;Set V0 to Key
  SKP V0	;Check Key (skip next instruction if pressed)
  JP checkup    ;Skip
  ADD V4,#1	;y++

  checkup:
  LD V0,#2	;Set V0 to Key
  SKP V0	;Check Key
  JP checkleft	;skip to next check
  ADD V4,#0xFF	;y--

  checkleft:
  LD V0,#4	;Set V0 to Key
  SKP V0	;Check Key
  JP checkright	;skip to next check
  ADD V3,#0xFF	;x--

  checkright:
  LD V0,#6	;Set V0 to Key
  SKP V0	;Check Key
  JP checkdone	;skip to next check
  ADD V3,#1	;x++

  checkdone:
  RET

INITENEMY:
  LD V5,#0
  RND V6,#15
  RET

DRAWENEMY:
  LD I,SPR_ENEMY	   ;Set I to SPR_ENEMY adress
  DRW V5,V6,#5
  RET

MOVEENEMY:
  ADD V5,#1

  ; compare V1>60
  LD V0,V5	
  LD V1,#60
  SUB V0,V1
  SNE VF,#1	
  CALL INITENEMY
  RET

WAITDELAY:
  LD V0,DT	   ;Set V0 to DelayTimer
  SE V0,#0	   ;Check zero
  JP WAITDELAY     
RET  

; ================== graphics ================
SPR_DUDE:
 DB #0b00111100
 DB #0b00011000
 DB #0b11111111
 DB #0b00011000
 DB #0b00100100
 DB #0b11100111

SPR_ENEMY:
 DB #0b01111110
 DB #0b11111111
 DB #0b10011001
 DB #0b11100111
 DB #0b00111100
	

code.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "yaca.h"
#include "code.h"

#define WRITE16(v) if (!error) fwrite(big2little16(v),sizeof(char)*2,1,myf)
#define WRITE8(v) if (!error) fwrite(v,sizeof(char)*1,1,myf)

#define ERROR_LINE(e) {fprintf(stderr,"Error - Line %d: ", e.line); error++;}

extern int _debug;

int error = 0;
static lex **mylab;
static FILE *myf;
static char *types[] = {"const value", "adress", "instruction", "register", "address"};


int *big2little16(int *opcode)
{
  int tmp = *opcode & 0xFF;
  *opcode>>=8;
  *opcode &= 0xFF;
  *opcode |= (tmp<<8);
}

int get_nb_ops(expr l)
{
  int num = 0;
  if (l.src != NULL) num++;
  if (l.dest != NULL) num++;
  if (l.op != NULL) num++;

  return num;
}

void code_scd(expr e)
{
  int opcode = 0x00C0;

  if (e.src->val > 0xF){
    ERROR_LINE(e);
    fprintf(stderr,"Constant value break 4bits limit\n");
    return;
  }
  opcode |= e.src->val;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_scr(expr e)
{
  int opcode = 0x00FB;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_low(expr e)
{
  int opcode = 0x00FE;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_high(expr e)
{
  int opcode = 0x00FF;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_scl(expr e)
{
  int opcode = 0x00FC;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

int search_in_lab(expr e)
{
  int i;

  for (i=0; mylab[i]!=NULL && i<50 && strcmp(mylab[i]->name,e.src->name); i++);
  
  if (i == 50 || mylab[i] == NULL){
    if (_debug)  printf("\n");
    ERROR_LINE(e);
    fprintf(stderr,"Label %s does not exist\n", e.src->name); 
    return ;//exit(BAD_LAB);
  }

  return mylab[i]->val;
}

void code_jmp(expr e)
{
  int opcode = (e.dest != NULL && e.dest->kind == REG)?0xB000:0x1000;
  
  
  opcode |= search_in_lab(e);
  if (_debug) printf("code: %x\n", opcode);
  WRITE16(&opcode);
}



void code_cls(expr e)
{
  int opcode = 0x00E0;
  if (_debug) printf("code: 00E0\n");
  WRITE16(&opcode);
}

void code_exit(expr e)
{
  int opcode = 0x00FD;
  if (_debug) printf("code: 00FD\n");
  WRITE16(&opcode);
}

void code_ld_r_c(expr e)
{
  int opcode = 0x6000;
  if (e.src->val > 0xFF){
    ERROR_LINE(e);
    fprintf(stderr,"Constant value break 1byte limit\n");
    return ;//exit(BAD_RANGE);
  }
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_r_r(expr e)
{
  int opcode = 0x8000;
 
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}


//TODO: gestion adresse directe
void code_ld_i_m(expr e)
{
  int opcode = 0xA000;
 
  opcode |=  search_in_lab(e);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}


void code_ld_r_dt(expr e)
{
  int opcode = 0xF007;
 
  opcode |=  e.dest->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_r_k(expr e)
{
  int opcode = 0xF00A;
 
  opcode |=  e.dest->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_dt_r(expr e)
{
  int opcode = 0xF015;

  opcode |=  e.src->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_st_r(expr e)
{
  int opcode = 0xF018;
 
  opcode |=  e.src->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_f_r(expr e)
{
  int opcode = 0xF029;
 
  opcode |=  e.src->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_b_r(expr e)
{
  int opcode = 0xF033;
 
  opcode |=  e.src->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_i_r(expr e)
{
  int opcode = 0xF055;
 
  opcode |=  e.src->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld_r_i(expr e)
{
  int opcode = 0xF065;
 
  opcode |=  e.src->val<<8;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ld(expr e)
{
  if (e.src->kind == VAL) code_ld_r_c(e);
  else if (e.src->kind == REG && e.dest->kind == REG) code_ld_r_r(e);
  else if (!strcmp(e.dest->name,"i")) code_ld_i_m(e);
  else if (!strcmp(e.src->name,"dt")) code_ld_r_dt(e);
  else if (!strcmp(e.src->name,"k")) code_ld_r_k(e);
  else if (!strcmp(e.dest->name,"dt")) code_ld_dt_r(e);
  else if (!strcmp(e.dest->name,"st")) code_ld_st_r(e);
  else if (!strcmp(e.dest->name,"f")) code_ld_f_r(e);
  else if (!strcmp(e.dest->name,"b")) code_ld_b_r(e);
  else if (!strcmp(e.dest->name,"[i]")) code_ld_i_r(e);
  else if (!strcmp(e.src->name,"[i]")) code_ld_r_i(e);
  else if (e.src->kind == ADDR){
    ERROR_LINE(e);
    fprintf(stderr,"%s %s,%s: ",e.inst->name, e.dest->name, e.src->name);
    fprintf(stderr,"src operand has address type\n");
  }
  else {
    ERROR_LINE(e);
    fprintf(stderr,"%s %s,%s is not recognize\n",e.inst->name, e.dest->name, e.src->name);
  }
}

void code_or(expr e)
{
  int opcode = 0x8001;
 
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_and(expr e)
{
  int opcode = 0x8002;
 
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_xor(expr e)
{
  int opcode = 0x8003;
 
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_sub(expr e)
{
  int opcode = 0x8005;
 
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_shr(expr e)
{
  int opcode = 0x8006;
  
  opcode |= (e.dest != NULL)?e.dest->val<<8:e.src->val<<8;
  opcode |= (e.dest != NULL)?e.src->val<<4:0;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_subn(expr e)
{
  int opcode = 0x8007;
  
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_shl(expr e)
{
  int opcode = 0x800E;
  
  opcode |= (e.dest->val<<8);
  opcode |= (e.src !=NULL)?(e.src->val<<4):0;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_call(expr e)
{
  int opcode = 0x2000;
  
  opcode |= search_in_lab(e);
  if (_debug) printf("code: %x\n", opcode);
  WRITE16(&opcode);
}

void code_rnd(expr e)
{
  int opcode = 0xC000;
  if (e.src->val > 0xFF){
    ERROR_LINE(e);
    fprintf(stderr,"Constant value break 1byte limit\n");
    return; //exit(BAD_RANGE);
  }
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val);
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_ret(expr e)
{
  int opcode = 0x00EE;
  if (_debug) printf("code: 00EE\n");
  WRITE16(&opcode);
}

void code_se_sne_c(expr e)
{
  int opcode = (!strcmp(e.inst->name,"se"))?0x3000:0x4000;
  
  if (e.src->val > 0xFF){
    ERROR_LINE(e);
    fprintf(stderr,"Constant value break 1byte limit\n");
    return ;//exit(BAD_RANGE);
  }
  
  opcode |= (e.dest->val<<8);
  opcode |= e.src->val;

  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_se(expr e)
{
  int opcode = 0x5000;
  
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);

  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_sne(expr e)
{
  int opcode = 0x9000;
  
  opcode |= (e.dest->val<<8);
  opcode |= (e.src->val<<4);

  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_se_sne(expr e)
{
  if (e.src->kind == VAL) code_se_sne_c(e);
  else if (e.src->kind == REG && !strcmp(e.inst->name, "se")) code_se(e);
  else if (e.src->kind == REG && !strcmp(e.inst->name, "sne")) code_sne(e);
  else if (e.dest->kind != REG){
    ERROR_LINE(e);
    fprintf(stderr,"dest operand has %s type and is not a register\n", types[e.dest->kind]);
    return ;//exit(BAD_OPS);
  }
  else{
    ERROR_LINE(e);
    fprintf(stderr,"src operand has %s type and is not a constant value\n", types[e.src->kind]);
    return ;
  }
}

void code_add_r_c(expr e)
{
   int opcode = 0x7000;
   
   if (e.src->val > 0xFF){
     ERROR_LINE(e);
     fprintf(stderr,"Constant value break 1byte limit\n");
     return ;//exit(BAD_RANGE);
   }
   
   opcode |= e.dest->val<<8;
   opcode |= e.src->val;
   if (_debug) printf("code: %X\n", opcode);
   WRITE16(&opcode);
}

void code_add_r_r(expr e)
{
   int opcode = 0x8004;
   
   opcode |= e.dest->val<<8;
   opcode |= e.src->val<<4;
   if (_debug) printf("code: %X\n", opcode);
   WRITE16(&opcode);
}

void code_add_i_r(expr e)
{
   int opcode = 0xF01E;
   
   opcode |= e.src->val<<8;
   if (_debug) printf("code: %X\n", opcode);
   WRITE16(&opcode);
}

void code_add(expr e)
{
  if (e.src->kind == VAL) code_add_r_c(e);
  else if (e.src->kind == REG && e.dest->kind == REG) code_add_r_r(e);
  else if (!strcmp(e.dest->name, "i")) code_add_i_r(e);
  else {
    ERROR_LINE(e);
    fprintf(stderr,"Unrecognize operand for %s\n", e.inst->name);
    return ;//exit(BAD_OPS);
  }
}

void code_drw(expr e)
{
  int opcode = 0xD000;

  if (e.op->kind != VAL){
    ERROR_LINE(e);
    fprintf(stderr,"Error: drw %s,%s,[%s] must have type const value but has type %s\n",e.src->name, e.dest->name, e.op->name, types[e.op->kind]);
    return ;//exit(BAD_VAL);
  }

  if (e.op->val > 0xF){
    ERROR_LINE(e);
    fprintf(stderr,"Constant value break 4bits limit\n");
    return ;//exit(BAD_RANGE);
  }
  
  opcode |= e.dest->val<<8;
  opcode |= e.src->val<<4;
  opcode |= e.op->val;
  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_skp_sknp(expr e)
{
  int opcode = (!strcmp(e.inst->name,"skp"))?0xE09E:0xE0A1;
  
  opcode |= (e.src->val<<8);

  if (_debug) printf("code: %X\n", opcode);
  WRITE16(&opcode);
}

void code_db_dw(expr e)
{
  int opcode = 0x0000;
  int is_db = !strcmp(e.inst->name,"db");
  if (e.src == NULL) return;
  opcode |= e.src->val;
  if (_debug) printf("code: %X\n", opcode);

  if (is_db && e.src->val > 0xFF){
    ERROR_LINE(e);
    fprintf(stderr,"Constant value break 1byte limit\n");
    return;
  } else if (!is_db && e.src->val > 0xFFFF){
    ERROR_LINE(e);
    fprintf(stderr,"Constant value break 2byte limit\n");
    return;
  }
  
  if (is_db) WRITE8(&opcode);
  else WRITE16(&opcode);
}



void debug_print_inst(expr l)
{
  int n = get_nb_ops(l);
  printf("%d: ",l.line);
  switch(n){
  case 0:
    printf("%s\t\t\t-> ", l.inst->name);
    break;
  case 1:
    printf("%s\t%s\t\t\t-> ", l.inst->name, l.src->name);
    break;
  case 2:
    printf("%s\t%s,%s\t-> ", l.inst->name, l.dest->name, l.src->name);
    break;
  case 3:
    printf("%s\t%s,%s,%s\t-> ", l.inst->name, l.dest->name, l.src->name, l.op->name);
  }
}

void code_prog(prog mprog, lex *lab[50], FILE *f)
{
  expr *tmp;
  mylab=lab;
  myf = f;
  
  for (tmp=mprog.first; tmp != NULL; tmp = tmp->next)
    {
      // NEED IMPROVEMENT FUNCTION POINTER TAB SHOULD DO THE WORK
      if (tmp->inst != NULL){
	if (_debug) debug_print_inst(*tmp);
	if (!strcmp(tmp->inst->name,"cls")) code_cls(*tmp);
	else if (!strcmp(tmp->inst->name,"ld")) code_ld(*tmp);
	else if (!strcmp(tmp->inst->name,"call")) code_call(*tmp);
	else if (!strcmp(tmp->inst->name,"ret")) code_ret(*tmp);
	else if (!strcmp(tmp->inst->name,"scd")) code_scd(*tmp);
	else if (!strcmp(tmp->inst->name,"scr")) code_scr(*tmp);
	else if (!strcmp(tmp->inst->name,"scl")) code_scl(*tmp);
	else if (!strcmp(tmp->inst->name,"low")) code_low(*tmp);
	else if (!strcmp(tmp->inst->name,"high")) code_high(*tmp);
	else if (!strcmp(tmp->inst->name,"jmp")) code_jmp(*tmp);
	else if (!strcmp(tmp->inst->name,"jp")) code_jmp(*tmp);
	else if (!strcmp(tmp->inst->name,"se")) code_se_sne(*tmp);
	else if (!strcmp(tmp->inst->name,"sne")) code_se_sne(*tmp);
	else if (!strncmp(tmp->inst->name,"add",3)) code_add(*tmp); // Ugly hack again
	else if (!strcmp(tmp->inst->name,"exit")) code_exit(*tmp);
	else if (!strcmp(tmp->inst->name,"or")) code_or(*tmp);
	else if (!strcmp(tmp->inst->name,"and")) code_and(*tmp);
	else if (!strcmp(tmp->inst->name,"xor")) code_xor(*tmp);
	else if (!strcmp(tmp->inst->name,"sub")) code_sub(*tmp);
	else if (!strcmp(tmp->inst->name,"shr")) code_shr(*tmp);
	else if (!strcmp(tmp->inst->name,"subn")) code_subn(*tmp);
	else if (!strcmp(tmp->inst->name,"shl")) code_shl(*tmp);
	else if (!strcmp(tmp->inst->name,"rnd")) code_rnd(*tmp);
	else if (!strcmp(tmp->inst->name,"drw")) code_drw(*tmp);
	else if (!strcmp(tmp->inst->name,"skp")) code_skp_sknp(*tmp);
	else if (!strcmp(tmp->inst->name,"sknp")) code_skp_sknp(*tmp);
	else if (!strcmp(tmp->inst->name,"db")) code_db_dw(*tmp);
	else if (!strcmp(tmp->inst->name,"dw")) code_db_dw(*tmp);
	else {
	  ERROR_LINE((*tmp));
	  fprintf(stderr,"Unrecognized %s instruction\n",tmp->inst->name);
	  //exit(BAD_OPS);
	}
      } 
    } 
}
